File: tests/Python/gateway.py
#!/usr/bin/env python3

# Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
#
# This file is part of DMSC.
# The DMSC project belongs to the Dunimd Team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
DMSC Gateway Module Python Tests.

This module contains comprehensive tests for the DMSC gateway system Python bindings.
The gateway system provides request routing, rate limiting, and circuit breaker
functionality for building resilient microservices.

Gateway Components:
- DMSCGateway: Main gateway entry point
- DMSCRouter: Request routing and rule management
- DMSCRoute: Individual route definition
- DMSCRateLimiter: Request rate control
- DMSCRateLimitConfig: Rate limiting configuration
- DMSCSlidingWindowRateLimiter: Sliding window rate limiting
- DMSCCircuitBreaker: Failure cascade prevention
- DMSCCircuitBreakerConfig: Circuit breaker configuration
- DMSCCircuitBreakerState: Circuit state enumeration

Request Routing:
- Route matching based on method and path
- Middleware support for cross-cutting concerns
- Handler delegation for business logic

Rate Limiting:
- Requests per second limits
- Burst capacity for traffic spikes
- Sliding window for smooth rate enforcement

Circuit Breaker Patterns:
- Closed: Normal operation
- Open: Blocking requests after failures
- Half-Open: Testing recovery

Test Classes:
- TestDMSCRouter: Router functionality
- TestDMSCGateway: Gateway entry point
- TestDMSCRoute: Route definition
- TestDMSCRateLimitConfig: Rate limit configuration
- TestDMSCCircuitBreakerConfig: Circuit breaker configuration
- TestDMSCCircuitBreakerState: State enumeration
- TestRateLimitStats: Rate limiting statistics
- TestCircuitBreakerMetrics: Circuit breaker metrics
"""

import unittest
from dmsc import (
    DMSCGateway, DMSCGatewayConfig, DMSCRouter, DMSCRoute,
    DMSCRateLimiter, DMSCRateLimitConfig, RateLimitStats,
    DMSCSlidingWindowRateLimiter, DMSCCircuitBreaker,
    DMSCCircuitBreakerConfig, DMSCCircuitBreakerState,
    CircuitBreakerMetrics
)


class TestDMSCRouter(unittest.TestCase):
    """
    Test suite for DMSCRouter class.

    The DMSCRouter class manages request routing rules and handles
    incoming requests by matching against registered routes. It
    provides efficient route lookup and management operations.

    Route Management:
    - add_route(): Register a new route
    - remove_route(): Unregister a route
    - clear_all_routes(): Remove all routes
    - get_route_count(): Number of registered routes

    Route Matching:
    - HTTP method matching (GET, POST, etc.)
    - Path pattern matching
    - Parameter extraction
    - Middleware chain execution

    Test Methods:
    - test_router_new: Verify router instantiation
    - test_router_get_route_count: Test route counting
    - test_router_clear_all_routes: Test route clearing
    """

    def test_router_new(self):
        """Test creating a new router.

        A router is created ready to accept route registrations
        and handle incoming requests.
        """
        router = DMSCRouter()
        self.assertIsNotNone(router)

    def test_router_get_route_count(self):
        """Test getting route count.
        
        The get_route_count() method returns the number of
        registered routes. A new router starts with zero routes.
        
        Expected Behavior:
        - Newly created router returns 0
        - After adding routes, count increases
        - After removing routes, count decreases
        - After clearing routes, count returns to 0
        """
        router = DMSCRouter()
        count = router.get_route_count()
        self.assertEqual(count, 0)

    def test_router_clear_all_routes(self):
        """Test clearing routes.
        
        The clear_all_routes() method removes all registered
        routes, resetting the router to its initial state.
        
        Expected Behavior:
        - After clear, route count is 0
        - Router can accept new route registrations
        - Previous route handlers are no longer accessible
        - Memory is freed for removed routes
        """
        router = DMSCRouter()
        router.clear_all_routes()
        self.assertEqual(router.get_route_count(), 0)


class TestDMSCGateway(unittest.TestCase):
    """Test suite for DMSCGateway class.
    
    The DMSCGateway class serves as the main entry point for the
    gateway system, coordinating routing, rate limiting, and
    circuit breaker functionality. It acts as a unified access
    point for all incoming requests.
    
    Gateway Responsibilities:
    - Request initialization: Parse incoming requests
    - Request routing: Match to appropriate handler
    - Rate limit enforcement: Check against limits before processing
    - Circuit breaker checks: Verify service availability
    - Response processing: Format and send responses
    - Metrics collection: Track request statistics
    
    Gateway Configuration:
    - Route table: Registered routes and handlers
    - Rate limiter: Token bucket or sliding window
    - Circuit breaker: Failure prevention mechanism
    - Middleware chain: Pre/post-processing hooks
    
    Request Flow:
    1. Request arrives at gateway
    2. Rate limiter checks (reject if over limit)
    3. Circuit breaker check (fail fast if open)
    4. Route matching (find handler)
    5. Middleware execution (pre-handler)
    6. Handler execution (business logic)
    7. Middleware execution (post-handler)
    8. Response sent to client
    
    Test Methods:
    - test_gateway_new: Verify gateway instantiation
    """

    def test_gateway_new(self):
        """Test creating a new gateway.
        
        This test verifies that DMSCGateway can be instantiated.
        The gateway is ready to be configured with routes and middleware.
        
        Expected Behavior:
        - Constructor completes without errors
        - Returns a valid gateway instance
        - Gateway has empty route table initially
        - Gateway is ready for configuration
        """
        gateway = DMSCGateway()
        self.assertIsNotNone(gateway)


class TestDMSCRoute(unittest.TestCase):
    """Test suite for DMSCRoute class.
    
    The DMSCRoute class defines a single routing rule including
    the HTTP method, path pattern, and associated handler. Routes
    are the fundamental building blocks of the gateway's routing
    system.
    
    Route Components:
    - Method: HTTP method (GET, POST, PUT, DELETE, PATCH, OPTIONS, HEAD)
    - Path: URL path pattern with optional parameters
    - Handler: Function or coroutine to process matching requests
    - Middleware: Pre/post-processing hooks for cross-cutting concerns
    - Metadata: Route tags, versioning, deprecation status
    
    Path Pattern Syntax:
    - Static: "/api/users" matches exactly "/api/users"
    - Parameter: "/api/users/{id}" captures id parameter
    - Optional: "/api/posts/{id?}" makes id optional
    - Wildcard: "/api/**" matches any path below
    
    Method Constraints:
    - GET: Retrieve resource (no body)
    - POST: Create new resource
    - PUT: Replace entire resource
    - PATCH: Partial update
    - DELETE: Remove resource
    
    Test Methods:
    - test_route_new: Verify route instantiation
    """

    def test_route_new(self):
        """Test creating a new route.
        
        This test verifies that DMSCRoute can be instantiated
        with a method and path pattern. The route is ready to
        be registered with a router.
        
        Expected Behavior:
        - Constructor accepts method and path parameters
        - Returns a valid route instance
        - Route is ready for handler assignment
        - Route can be added to router
        """
        route = DMSCRoute("GET", "/test")
        self.assertIsNotNone(route)


class TestDMSCRateLimitConfig(unittest.TestCase):
    """Test suite for DMSCRateLimitConfig class.
    
    The DMSCRateLimitConfig class configures rate limiting behavior
    including request rates, burst capacities, and policy enforcement.
    Rate limiting protects services from overuse and ensures fair access.
    
    Rate Limit Parameters:
    - requests_per_second: Maximum average request rate (sustained limit)
    - burst_size: Maximum burst capacity above average rate (peak tolerance)
    - window_size: Time window for rate calculation (default 1 second)
    - policy: Enforcement policy (reject, queue, or delay)
    
    Rate Limiting Algorithms:
    - Token Bucket: Tokens added at fixed rate, consumed per request
    - Leaky Bucket: Requests queued, processed at fixed rate
    - Sliding Window: Moving time window for smooth limits
    - Fixed Window: Simple counting within time windows
    
    Configuration Examples:
    - API Limit: 100 requests/second, 200 burst
    - WebSocket: 50 messages/second, 100 burst
    - File Upload: 10 uploads/second, 20 burst
    
    Test Methods:
    - test_rate_limit_config_setters: Test configuration updates
    """

    def test_rate_limit_config_setters(self):
        """Test rate limit configuration setters.
        
        Rate limit parameters can be configured to protect different
        endpoints based on their expected load:
        - 50 requests per second average rate
        - 100 requests burst capacity for traffic spikes
        
        Expected Behavior:
        - requests_per_second setter updates the rate
        - burst_size setter updates the burst capacity
        - get_requests_per_second returns configured value
        - get_burst_size returns configured value
        """
        config = DMSCRateLimitConfig()
        config.set_requests_per_second(50)
        config.set_burst_size(100)
        self.assertEqual(config.get_requests_per_second(), 50)
        self.assertEqual(config.get_burst_size(), 100)


class TestDMSCCircuitBreakerConfig(unittest.TestCase):
    """Test suite for DMSCCircuitBreakerConfig class.
    
    The DMSCCircuitBreakerConfig class configures circuit breaker
    behavior including failure thresholds, timeout periods, and
    recovery settings. Circuit breakers prevent cascading failures
    in distributed systems.
    
    Circuit Breaker Parameters:
    - failure_threshold: Number of failures before opening circuit
    - success_threshold: Successes needed in half-open to close
    - timeout_seconds: Time in open state before half-open transition
    - volume_threshold: Minimum requests before evaluating failure rate
    
    State Transitions:
    - Closed -> Open: Exceeds failure threshold or error rate
    - Open -> Half-Open: Timeout expires, testing recovery
    - Half-Open -> Closed: Sufficient successes (meets threshold)
    - Half-Open -> Open: Any failure during testing
    
    Failure Detection:
    - Count-based: N consecutive failures
    - Rate-based: Error rate exceeds percentage
    - Volume-based: Minimum sample size required
    
    Test Methods:
    - test_circuit_breaker_config_setters: Test configuration updates
    """

    def test_circuit_breaker_config_setters(self):
        """Test circuit breaker configuration setters.
        
        Circuit breaker thresholds control sensitivity to failures
        and recovery behavior:
        - Open after 10 consecutive failures
        - Close after 3 successes in half-open state
        
        Expected Behavior:
        - failure_threshold setter updates the failure limit
        - success_threshold setter updates the recovery threshold
        - get_failure_threshold returns configured value
        - get_success_threshold returns configured value
        """
        config = DMSCCircuitBreakerConfig()
        config.set_failure_threshold(10)
        config.set_success_threshold(3)
        self.assertEqual(config.get_failure_threshold(), 10)
        self.assertEqual(config.get_success_threshold(), 3)


class TestDMSCCircuitBreakerState(unittest.TestCase):
    """Test suite for DMSCCircuitBreakerState enum.
    
    The DMSCCircuitBreakerState enum represents the current state
    of a circuit breaker, determining how requests are handled.
    States implement the fail-fast pattern to protect failing services.
    
    Circuit States:
    - Closed: Normal operation, requests pass through
    - Open: Requests fail immediately without attempting service
    - Half-Open: Limited requests allowed to test recovery
    
    State Characteristics:
    - Closed: Requests execute normally, failures counted
    - Open: Requests return error immediately, no execution
    - Half-Open: Single request allowed to test service
    
    State Purpose:
    States prevent cascading failures by stopping requests to
    failing services, allowing them time to recover, and
    gradually restoring traffic as health returns.
    
    State Duration:
    - Closed: Indefinite (until failure threshold exceeded)
    - Open: Until timeout expires, then transitions to Half-Open
    - Half-Open: Until success threshold met or failure occurs
    
    Test Methods:
    - test_circuit_breaker_state_values: Verify state values
    """

    def test_circuit_breaker_state_values(self):
        """Test circuit breaker state values.
        
        Each circuit state should have a string representation
        for logging, monitoring, and debugging purposes.
        
        Expected Behavior:
        - Closed state string matches expected format
        - Open state string matches expected format
        - Half-Open state string matches expected format
        """
        self.assertEqual(str(DMSCCircuitBreakerState.Closed), "DMSCCircuitBreakerState.Closed")
        self.assertEqual(str(DMSCCircuitBreakerState.Open), "DMSCCircuitBreakerState.Open")
        self.assertEqual(str(DMSCCircuitBreakerState.HalfOpen), "DMSCCircuitBreakerState.HalfOpen")


class TestRateLimitStats(unittest.TestCase):
    """Test suite for RateLimitStats class.
    
    The RateLimitStats class provides statistics about rate limiting
    including current token bucket state, request counts, and limit
    enforcement metrics. These statistics are essential for monitoring
    and alerting on API usage.
    
    Statistics Tracked:
    - current_tokens: Remaining tokens in the bucket (0 to max)
    - total_requests: Total requests processed since start
    - rejected_requests: Number of requests rejected (over limit)
    - queued_requests: Number of requests queued (if queuing enabled)
    - last_refill: Timestamp of last token refill
    
    Metrics Usage:
    - Monitor rate limit effectiveness: Track rejection rates
    - Detect traffic anomalies: Sudden changes in patterns
    - Plan capacity requirements: Understand peak usage
    - Optimize limits: Adjust based on actual usage
    
    Performance Monitoring:
    - Hit rate: Requests served vs total requests
    - Rejection rate: Rejected vs total requests
    - Token utilization: Average tokens consumed
    
    Test Methods:
    - test_rate_limit_stats_new: Verify stats instantiation
    - test_rate_limit_stats_properties: Test property access
    """

    def test_rate_limit_stats_new(self):
        """Test creating rate limit stats.
        
        Rate limit statistics track the state of rate limiting
        for monitoring, analysis, and alerting purposes.
        
        Expected Behavior:
        - Constructor accepts initial token count and request count
        - Returns a valid stats instance
        - Stats object tracks rate limiting metrics
        """
        stats = RateLimitStats(100, 50)
        self.assertIsNotNone(stats)

    def test_rate_limit_stats_properties(self):
        """Test rate limit stats properties.
        
        Statistics provide access to current rate limiting state:
        - 100 tokens available in bucket
        - 50 total requests processed
        
        Expected Behavior:
        - get_current_tokens() returns initial token count
        - get_total_requests() returns initial request count
        - Properties can be updated during operation
        """
        stats = RateLimitStats(100, 50)
        self.assertEqual(stats.get_current_tokens(), 100)
        self.assertEqual(stats.get_total_requests(), 50)


class TestCircuitBreakerMetrics(unittest.TestCase):
    """Test suite for CircuitBreakerMetrics class.
    
    The CircuitBreakerMetrics class tracks circuit breaker performance
    including state transitions, request counts, and failure rates.
    These metrics are crucial for understanding service health and
    circuit breaker effectiveness.
    
    Metrics Tracked:
    - current_state: Current circuit breaker state
    - failure_count: Number of recent failures
    - success_count: Number of recent successes
    - request_count: Total requests attempted
    - rejected_count: Number of requests rejected (when open)
    
    Performance Analysis:
    - Failure rate: failures / (failures + successes)
    - State duration: Time spent in each state
    - Recovery time: How long until circuit closes
    - Protection effectiveness: Requests rejected vs failed
    
    Alerting Thresholds:
    - Circuit open for extended period
    - High failure rate sustained
    - Frequent state transitions
    
    Test Methods:
    - test_circuit_breaker_metrics_new: Verify metrics instantiation
    """

    def test_circuit_breaker_metrics_new(self):
        """Test creating circuit breaker metrics.
        
        Circuit breaker metrics track performance for monitoring,
        alerting, and capacity planning. They help identify
        problematic services and tune circuit breaker thresholds.
        
        Expected Behavior:
        - Constructor accepts state and count parameters
        - Returns a valid metrics instance
        - Metrics object tracks circuit breaker performance
        """
        metrics = CircuitBreakerMetrics(
            "Closed", 0, 0, 0, 0
        )
        self.assertIsNotNone(metrics)


if __name__ == "__main__":
    unittest.main()

File: tests/Rust/gateway.rs
//! Copyright © 2025-2026 Wenze Wei. All Rights Reserved.
//!
//! This file is part of Ri.
//! The Ri project belongs to the Dunimd Team.
//!
//! Licensed under the Apache License, Version 2.0 (the "License");
//! You may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//!
//!     http://www.apache.org/licenses/LICENSE-2.0
//!
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

//! # Gateway Module Tests
//!
//! This module contains comprehensive tests for the Ri API gateway system,
//! covering request/response handling, routing, middleware processing, load balancing,
//! rate limiting, and circuit breaker functionality for building resilient API services.
//!
//! ## Test Coverage
//!
//! - **RiGatewayConfig**: Tests for gateway configuration including network settings
//!   (listen address, port, max connections), feature toggles (rate limiting, circuit
//!   breaker, load balancing), CORS settings, and logging configuration
//!
//! - **RiGatewayRequest**: Tests for request object creation and properties including
//!   HTTP method, path, headers, query parameters, body, and remote address
//!
//! - **RiGatewayResponse**: Tests for response construction including status codes,
//!   headers, body content, JSON responses, and error responses
//!
//! - **RiRouter**: Tests for route registration, route matching, and route counting
//!
//! - **RiMiddleware**: Tests for middleware chain execution and the interceptor pattern
//!   for request/response processing
//!
//! - **RiLoadBalancer**: Tests for load balancing strategies (RoundRobin, LeastConn,
//!   Random), backend server management, and healthy server selection
//!
//! - **RiRateLimiter**: Tests for rate limiting configuration and request throttling
//!
//! - **RiCircuitBreaker**: Tests for circuit breaker states (Closed, Open, Half-Open)
//!   and fault tolerance patterns
//!
//! ## Architecture
//!
//! The gateway system implements a layered architecture:
//! - **Transport Layer**: Handles raw network connections and request parsing
//! - **Router Layer**: Matches incoming requests to registered routes using method and path
//! - **Middleware Layer**: Applies cross-cutting concerns (authentication, logging, etc.)
//! - **Handler Layer**: Executes business logic for matched routes
//! - **Response Layer**: Formats and sends responses back to clients
//!
//! ## Load Balancing Strategies
//!
//! The load balancer supports multiple distribution strategies:
//! - **RoundRobin**: Sequentially distributes requests across all healthy backends
//! - **LeastConn**: Routes to the backend with the fewest active connections
//! - **Random**: Randomly selects a healthy backend for each request
//!
//! ## Circuit Breaker States
//!
//! The circuit breaker implements a three-state failure protection pattern:
//! - **Closed**: Normal operation, requests pass through to backends
//! - **Open**: After threshold failures, requests are immediately rejected
//! - **Half-Open**: Probe requests are allowed to test backend recovery
//!
//! ## Middleware Chain
//!
//! The middleware chain enables flexible request processing composition:
//! - Middlewares are registered in order and execute sequentially
//! - Each middleware can modify request headers, body, or reject the request
//! - Errors in middleware short-circuit the chain and return error responses
//! - The chain pattern enables separation of concerns (auth, logging, compression, etc.)

use ri::gateway::{
    RiGatewayConfig,
    RiGatewayRequest,
    RiGatewayResponse,
    RiGateway,
    RiRoute,
    RiRouter,
    RiMiddleware,
    RiMiddlewareChain,
    RiLoadBalancer,
    RiLoadBalancerStrategy,
    RiRateLimiter,
    RiRateLimitConfig,
    RiCircuitBreaker,
    RiCircuitBreakerConfig,
};
use ri::gateway::load_balancer::RiBackendServer;
use ri::prelude::RiResult;
use std::collections::HashMap;
use std::sync::Arc;

#[test]
/// Tests RiGatewayConfig default configuration values.
///
/// Verifies that the default gateway configuration has appropriate values
/// for network settings, feature toggles, CORS, and logging.
///
/// ## Default Configuration Values
///
/// - listen_address: "0.0.0.0" - Listen on all network interfaces
/// - listen_port: 8080 - Default HTTP port
/// - max_connections: 10000 - Maximum concurrent connections
/// - request_timeout_seconds: 30 - Request timeout
/// - enable_rate_limiting: true - Rate limiting enabled
/// - enable_circuit_breaker: true - Circuit breaker enabled
/// - enable_load_balancing: true - Load balancing enabled
/// - cors_enabled: true - CORS support enabled
/// - cors_origins: ["*"] - Allow all origins
/// - enable_logging: true - Request logging enabled
/// - log_level: "info" - Log level set to info
///
/// ## Expected Behavior
///
/// All configuration fields have sensible defaults suitable for
/// typical API gateway deployments.
fn test_gateway_config_default() {
    let config = RiGatewayConfig::default();
    
    assert_eq!(config.listen_address, "0.0.0.0");
    assert_eq!(config.listen_port, 8080);
    assert_eq!(config.max_connections, 10000);
    assert_eq!(config.request_timeout_seconds, 30);
    assert!(config.enable_rate_limiting);
    assert!(config.enable_circuit_breaker);
    assert!(config.enable_load_balancing);
    assert!(config.cors_enabled);
    assert_eq!(config.cors_origins, vec!["*".to_string()]);
    assert!(config.enable_logging);
    assert_eq!(config.log_level, "info");
}

#[test]
/// Tests RiGatewayRequest creation and properties.
///
/// Verifies that a gateway request can be created with all required
/// properties including HTTP method, path, headers, query parameters,
/// body, and remote address.
///
/// ## Request Properties
///
/// - **id**: Unique identifier generated automatically
/// - **method**: HTTP method (GET, POST, etc.)
/// - **path**: Request path
/// - **headers**: HTTP headers as key-value pairs
/// - **query_params**: Query string parameters
/// - **body**: Optional request body as bytes
/// - **remote_addr**: Client address in format "IP:port"
///
/// ## Expected Behavior
///
/// - A unique request ID is generated
/// - All properties are stored correctly
/// - The request is ready for processing
fn test_gateway_request_new() {
    let method = "GET".to_string();
    let path = "/test".to_string();
    let headers = std::collections::HashMap::new();
    let query_params = std::collections::HashMap::new();
    let body = None::<Vec<u8>>;
    let remote_addr = "127.0.0.1:12345".to_string();
    
    let request = RiGatewayRequest::new(
        method.clone(),
        path.clone(),
        headers.clone(),
        query_params.clone(),
        body.clone(),
        remote_addr.clone(),
    );
    
    assert!(!request.id.is_empty());
    assert_eq!(request.method, method);
    assert_eq!(request.path, path);
    assert_eq!(request.headers, headers);
    assert_eq!(request.query_params, query_params);
    assert_eq!(request.body, body);
    assert_eq!(request.remote_addr, remote_addr);
}

#[test]
/// Tests RiGatewayResponse creation with status, body, and request ID.
///
/// Verifies that a gateway response can be created with the essential
/// properties and that default headers are added automatically.
///
/// ## Default Headers
///
/// - **Content-Type**: Set based on response type
/// - **X-Request-ID**: Copied from request for tracing
///
/// ## Expected Behavior
///
/// - Response is created with specified values
/// - Default headers are present
/// - The response is ready to be sent
fn test_gateway_response_new() {
    let status_code = 200;
    let body = b"test_body".to_vec();
    let request_id = "test_request_id".to_string();
    
    let response = RiGatewayResponse::new(status_code, body.clone(), request_id.clone());
    
    assert_eq!(response.status_code, status_code);
    assert_eq!(response.body, body);
    assert_eq!(response.request_id, request_id);
    assert!(response.headers.contains_key("Content-Type"));
    assert!(response.headers.contains_key("X-Request-ID"));
}

#[test]
/// Tests RiGatewayResponse header addition with with_header().
///
/// Verifies that custom headers can be added to responses using
/// the builder pattern through with_header() method.
///
/// ## Expected Behavior
///
/// - Custom headers can be added
/// - Multiple headers can be added via chaining
/// - Added headers are accessible via get()
fn test_gateway_response_with_header() {
    let status_code = 200;
    let body = b"test_body".to_vec();
    let request_id = "test_request_id".to_string();
    
    let response = RiGatewayResponse::new(status_code, body.clone(), request_id.clone())
        .with_header("Custom-Header".to_string(), "Custom-Value".to_string());
    
    assert_eq!(response.headers.get("Custom-Header"), Some(&"Custom-Value".to_string()));
}

#[test]
/// Tests RiGatewayResponse JSON creation with json() factory.
///
/// Verifies that JSON responses can be created easily using the
/// json() factory method which sets the Content-Type header.
///
/// ## JSON Response Behavior
///
/// - Serializes the data to JSON
/// - Sets Content-Type to application/json
/// - Includes the request ID for tracing
///
/// ## Expected Behavior
///
/// - Response contains serialized JSON
/// - Content-Type is set to application/json
/// - Status code and request ID are preserved
fn test_gateway_response_json() {
    let status_code = 200;
    let request_id = "test_request_id".to_string();
    
    let data = serde_json::json!({"key": "value"});
    
    let response = RiGatewayResponse::json(status_code, &data, request_id.clone()).unwrap();
    
    assert_eq!(response.status_code, status_code);
    assert_eq!(response.request_id, request_id);
    assert!(response.headers.contains_key("Content-Type"));
    assert_eq!(response.headers.get("Content-Type"), Some(&"application/json".to_string()));
}

#[test]
/// Tests RiGatewayResponse error creation with error() factory.
///
/// Verifies that error responses can be created easily using the
/// error() factory method for standardized error handling.
///
/// ## Error Response Behavior
///
/// - Sets appropriate status code
/// - Includes error message in body
/// - Sets Content-Type to text/plain
///
/// ## Expected Behavior
///
/// - Error response contains message
/// - Status code matches requested code
/// - Headers are set appropriately
fn test_gateway_response_error() {
    let status_code = 404;
    let message = "Not Found".to_string();
    let request_id = "test_request_id".to_string();
    
    let response = RiGatewayResponse::error(status_code, message.clone(), request_id.clone());
    
    assert_eq!(response.status_code, status_code);
    assert_eq!(response.request_id, request_id);
    assert!(response.headers.contains_key("Content-Type"));
    assert!(response.body.len() > 0);
}

#[tokio::test]
/// Tests RiGateway creation and initial state.
///
/// Verifies that a gateway can be created successfully and starts
/// with empty routes and middleware chain.
///
/// ## Initial Gateway State
///
/// - Router has zero routes registered
/// - Middleware chain is empty
/// - Gateway is ready for configuration
///
/// ## Expected Behavior
///
/// - Gateway is created without errors
/// - Initial state is as expected
async fn test_gateway_new() {
    let gateway = RiGateway::new();
    
    // Verify gateway components are created
    assert_eq!(gateway.router().route_count(), 0);
    assert_eq!(gateway.middleware_chain().len(), 0);
}

#[tokio::test]
/// Tests RiRouter route registration with add_route().
///
/// Verifies that routes can be registered with the router and that
/// the route count is updated accordingly.
///
/// ## Route Registration
///
/// - Routes are registered with method, path, and handler
/// - Each route gets a unique identifier
/// - Route count reflects registered routes
///
/// ## Expected Behavior
///
/// - Route is successfully registered
/// - Route count increases by 1
/// - Route is available for matching
async fn test_gateway_router() {
    let router = RiRouter::new();
    
    // Test adding a route
    let handler = Arc::new(|request: RiGatewayRequest| {
        Box::pin(async move {
            Ok(RiGatewayResponse::new(
                200,
                b"test_response".to_vec(),
                request.id,
            ))
        }) as std::pin::Pin<Box<dyn std::future::Future<Output = RiResult<RiGatewayResponse>> + Send>>
    });

    let route = RiRoute::new(
        "GET".to_string(),
        "/test".to_string(),
        handler,
    );

    router.add_route(route);

    assert_eq!(router.route_count(), 1);
}

#[tokio::test]
/// Tests RiMiddlewareChain middleware execution.
///
/// Verifies that middlewares can be added to the chain and that
/// they are executed in order when processing requests.
///
/// ## Middleware Execution
///
/// - Middlewares are added in order
/// - Each middleware can modify the request
/// - Execution proceeds through the chain
///
/// ## Expected Behavior
///
/// - Middleware is added to the chain
/// - Middleware execute() is called
/// - Request modifications are applied
async fn test_gateway_middleware_chain() {
    struct TestMiddleware;

    #[async_trait::async_trait]
    impl RiMiddleware for TestMiddleware {
        async fn execute(
            &self,
            request: &mut RiGatewayRequest,
        ) -> RiResult<()> {
            request
                .headers
                .insert("X-Custom-Middleware".to_string(), "applied".to_string());
            Ok(())
        }

        fn name(&self) -> &'static str {
            "TestMiddleware"
        }
    }

    let mut middleware_chain = RiMiddlewareChain::new();

    middleware_chain.add(Arc::new(TestMiddleware));

    let mut request = RiGatewayRequest::new(
        "GET".to_string(),
        "/test".to_string(),
        HashMap::new(),
        HashMap::new(),
        None,
        "127.0.0.1:12345".to_string(),
    );

    middleware_chain.execute(&mut request).await.unwrap();

    assert_eq!(
        request.headers.get("X-Custom-Middleware"),
        Some(&"applied".to_string())
    );
    assert_eq!(middleware_chain.len(), 1);
}

#[tokio::test]
/// Tests RiGateway request handling with handle_request().
///
/// Verifies that the gateway can handle requests through the
/// complete request-response cycle including routing and middleware.
///
/// ## Request Flow
///
/// 1. Request is received by gateway
/// 2. Middleware chain processes request
/// 3. Router matches request to route
/// 4. Handler processes request
/// 5. Response is returned to client
///
/// ## Expected Behavior
///
/// - Request is routed to matching handler
/// - Handler response is returned
/// - Status code matches expected value
async fn test_gateway_handle_request() {
    let gateway = RiGateway::new();
    let router = gateway.router();
    
    // Add a test route
    let handler = Arc::new(|request: RiGatewayRequest| {
        Box::pin(async move {
            Ok(RiGatewayResponse::new(
                200,
                b"test_response".to_vec(),
                request.id,
            ))
        }) as std::pin::Pin<Box<dyn std::future::Future<Output = RiResult<RiGatewayResponse>> + Send>>
    });

    let route = RiRoute::new(
        "GET".to_string(),
        "/test".to_string(),
        handler,
    );

    router.add_route(route);
    
    // Create a test request
    let request = RiGatewayRequest::new(
        "GET".to_string(),
        "/test".to_string(),
        std::collections::HashMap::new(),
        std::collections::HashMap::new(),
        None,
        "127.0.0.1:12345".to_string(),
    );
    
    // Handle the request
    let response = gateway.handle_request(request).await;
    
    // Verify the response
    assert_eq!(response.status_code, 200);
    assert_eq!(response.body, b"test_response".to_vec());
}

#[tokio::test]
/// Tests RiGateway 404 response for unmatched routes.
///
/// Verifies that the gateway returns a 404 status code when
/// no matching route is found for the request path.
///
/// ## 404 Handling
///
/// - Unmatched requests return 404
/// - Error response includes message
/// - Request ID is preserved for tracing
///
/// ## Expected Behavior
///
/// - Non-existent route returns 404
/// - Error message is included
async fn test_gateway_handle_request_not_found() {
    let gateway = RiGateway::new();
    
    // Create a test request to a non-existent route
    let request = RiGatewayRequest::new(
        "GET".to_string(),
        "/non_existent_route".to_string(),
        std::collections::HashMap::new(),
        std::collections::HashMap::new(),
        None,
        "127.0.0.1:12345".to_string(),
    );
    
    // Handle the request
    let response = gateway.handle_request(request).await;
    
    // Verify the response is 404
    assert_eq!(response.status_code, 404);
}

#[tokio::test]
/// Tests RiLoadBalancer server management and selection.
///
/// Verifies that the load balancer can manage backend servers
/// and select healthy servers using the configured strategy.
///
/// ## Load Balancer Operations
///
/// - Servers can be added to the pool
/// - Healthy servers can be retrieved
/// - Server selection uses strategy
///
/// ## Expected Behavior
///
/// - Servers are added successfully
/// - All servers are retrieved as healthy
/// - Selected server matches pattern
async fn test_load_balancer_new() {
    let load_balancer = RiLoadBalancer::new(RiLoadBalancerStrategy::RoundRobin);
    
    // Test adding targets
    let server1 = RiBackendServer::new(
        "server1".to_string(),
        "http://localhost:8001".to_string(),
    );
    let server2 = RiBackendServer::new(
        "server2".to_string(),
        "http://localhost:8002".to_string(),
    );

    load_balancer.add_server(server1).await;
    load_balancer.add_server(server2).await;
    
    // Test getting targets
    let targets = load_balancer.get_healthy_servers().await;
    assert_eq!(targets.len(), 2);
    
    // Test selecting a target
    let target = load_balancer.select_server(None).await.unwrap();
    assert!(target.url.starts_with("http://localhost"));
}

#[tokio::test]
/// Tests RiRateLimiter request throttling.
///
/// Verifies that the rate limiter can check if requests should
/// be allowed based on the configured rate limit settings.
///
/// ## Rate Limiting Behavior
///
/// - Requests within limit are allowed
/// - Requests exceeding limit are rejected
/// - Limits are per configuration window
///
/// ## Expected Behavior
///
/// - Initial request is allowed
/// - Rate limiting operates correctly
async fn test_rate_limiter_new() {
    let config = RiRateLimitConfig {
        requests_per_second: 100,
        burst_size: 200,
        window_seconds: 1,
    };
    
    let rate_limiter = RiRateLimiter::new(config);
    
    // Test checking a request
    let request = RiGatewayRequest::new(
        "GET".to_string(),
        "/test".to_string(),
        std::collections::HashMap::new(),
        std::collections::HashMap::new(),
        None,
        "127.0.0.1:12345".to_string(),
    );
    
    let allowed = rate_limiter.check_request(&request).await;
    assert!(allowed);
}

#[tokio::test]
/// Tests RiCircuitBreaker state management.
///
/// Verifies that the circuit breaker tracks success and failure
/// counts and allows requests in the appropriate states.
///
/// ## Circuit Breaker States
///
/// - **Closed**: Normal operation, requests allowed
/// - **Open**: Too many failures, requests rejected
/// - **Half-Open**: Testing recovery, limited requests
///
/// ## Expected Behavior
///
/// - Initial state allows requests
/// - Success/failure recording works
async fn test_circuit_breaker_new() {
    let config = RiCircuitBreakerConfig {
        failure_threshold: 5,
        success_threshold: 3,
        timeout_seconds: 30,
        monitoring_period_seconds: 30,
    };
    
    let circuit_breaker = RiCircuitBreaker::new(config);
    
    // Test allowing requests initially
    let allowed = circuit_breaker.allow_request().await;
    assert!(allowed);
    
    // Test recording success
    circuit_breaker.record_success().await;
    
    // Test recording failure
    circuit_breaker.record_failure().await;
}
